src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,432 @@
+            var binaryExpressionSyntaxKind = GenerateBinaryExpressionSyntaxKind(syntaxNode);
+            rewrittenSyntaxNode = SyntaxFactory.BinaryExpression(binaryExpressionSyntaxKind, rewrittenAssignmentExpression.Left, rewrittenAssignmentExpression.Right);
+        }
+
+        var rewrittenBuildersExpression = rewrittenSyntaxNode switch
+        {
+            AssignmentExpressionSyntax @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => rewrittenSyntaxNode
+        };
+
+        var (finalBuildersExpression, finalConstantMapper) = ConstantsMapperComposer.Compose(rewrittenBuildersExpression, builderNodeToVariableValue.Values.Select(v => v.ExpressionAnalysisContext.Node.ConstantsRemapper));
+        return new ExpressionAnalysisContext(new ExpressionAnalysisNode(rewrittenBuildersExpression, argumentTypeName, finalBuildersExpression, finalConstantMapper));
+    }
+
+    private static bool ProcessDeclaration(VariableDeclaratorSyntax variableDeclarator, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var evaluationResult = PreEvaluate(variableDeclarator, contextInformation, null, displayDiagnostics);
+        var variableName = variableDeclarator.Identifier.ValueText;
+
+        if (evaluationResult.CanEvaluate)
+        {
+            var variableValue = evaluationResult.SyntaxNodeToVariableValueMap.Values.First();
+            var generateNewContext = evaluationResult.SyntaxNodeToVariableValueMap.Count() > 1;
+
+            if (generateNewContext)
+            {
+                var expressionAnalysisContext = GenerateNewExpressionAnalysisContext(variableDeclarator, evaluationResult.SyntaxNodeToVariableValueMap);
+                variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+            }
+
+            dirtyVariables.Add(variableName);
+            contextInformation.CurrentVariablesValues.Add(variableName, variableValue);
+
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.ProcessedVariables.Add(variableValue);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(variableName, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static bool ProcessAssignment(AssignmentExpressionSyntax assignmentExpression, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var evaluationResult = PreEvaluate(assignmentExpression, contextInformation, null, displayDiagnostics);
+        var variableName = (assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax).Identifier.ValueText;
```

Comment:
Comment for why Left is IdentifierNameSyntax

Reply:
Done